File: src/leetcode/java/Candy_135.java
package leetcode.java.hard;

public class Candy_135 {
  public int candy(int[] ratings) {
    int[] values = new int[ratings.length];
    int[] candies = new int[ratings.length];
    candies[0] = 1;
    for (int i = 1; i < ratings.length; i++) {
      values[i] = ratings[i] - ratings[i - 1];
      candies[i] = 1;
    }

    for (int j = ratings.length - 1; j >= 1; j--) {
      if (values[j] < 0 && candies[j - 1] <= candies[j]) {
        candies[j - 1] = candies[j] + 1;
      }
    }
    int totalCandies = 0;
    for (int k = 1; k < ratings.length; k++) {
      if (values[k] > 0 && candies[k - 1] >= candies[k]) {
        candies[k] = candies[k - 1] + 1;
      }
      totalCandies += candies[k - 1];
    }
    totalCandies += candies[ratings.length - 1];
    return totalCandies;
  }
}

File: src/leetcode/python/Candy_135.py
class Candy_135(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        length = len(ratings)
        candies = [1] * length

        for j in range(length - 1, 0, -1):
            if ratings[j] - ratings[j - 1] < 0 and candies[j - 1] <= candies[j]:
                candies[j - 1] = candies[j] + 1
        for k in range(1, length):
            if ratings[k] - ratings[k - 1] > 0 and candies[k - 1] >= candies[k]:
                candies[k] = candies[k - 1] + 1
        return sum(candies)
